File: benchmark/cases/oop_dispatch/oop_dispatch.js
class Base {
    compute() { return 0; }
}

class TypeA extends Base {
    compute() { return 1; }
}

class TypeB extends Base {
    compute() { return 2; }
}

class TypeC extends Base {
    compute() { return 3; }
}

class TypeD extends Base {
    compute() { return 4; }
}

const classes = [TypeA, TypeB, TypeC, TypeD];
let sum = 0;
for (let i = 0; i < 100000; i++) {
    const obj = new classes[i % 4]();
    sum += obj.compute();
}
console.log(sum);

File: benchmark/cases/oop_dispatch/oop_dispatch.lua
local Base = {}
Base.__index = Base
function Base:new() return setmetatable({}, self) end
function Base:compute() return 0 end

local TypeA = setmetatable({}, {__index = Base})
TypeA.__index = TypeA
function TypeA:compute() return 1 end

local TypeB = setmetatable({}, {__index = Base})
TypeB.__index = TypeB
function TypeB:compute() return 2 end

local TypeC = setmetatable({}, {__index = Base})
TypeC.__index = TypeC
function TypeC:compute() return 3 end

local TypeD = setmetatable({}, {__index = Base})
TypeD.__index = TypeD
function TypeD:compute() return 4 end

local classes = {TypeA, TypeB, TypeC, TypeD}
local sum = 0
for i = 0, 99999 do
    local cls = classes[(i % 4) + 1]
    local obj = cls:new()
    sum = sum + obj:compute()
end
print(sum)
